{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyNe3q+5Z+QuDlJytvLVYJtV"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["# Analyze human cancer data from The Cancer Genome Atlas via LinkedOmics\n","### Data freely available here: https://www.linkedomics.org/login.php"],"metadata":{"id":"KAvbaQnrDH0c"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"ltM4j96I-xRv"},"outputs":[],"source":["%pip install -q ipss\n","\n","import os\n","\n","from ipss import ipss\n","import matplotlib.pyplot as plt\n","import numpy as np\n","import pandas as pd\n","import urllib.request"]},{"cell_type":"markdown","source":["## Function for loading the cancer data (do not alter)"],"metadata":{"id":"2-7zk7zKEZ6d"}},{"cell_type":"code","source":["# Dictionary of direct download links from Google Drive for both ovarian and prostate files\n","file_urls = {\n","    'ovarian': {\n","        'clinical': 'https://drive.google.com/uc?export=download&id=1rL47G7dVaaQQrrntKLr3hkZKe_ZNmSap',\n","        'methylation': 'https://drive.google.com/uc?export=download&id=11sRNgtLSYDCD6hne32XLHkQd1PF-Vpl_',\n","        'mirna': 'https://drive.google.com/uc?export=download&id=1lc-If-Q7icU7p5N_lWouFrkqf_IJEuSd',\n","        'rnaseq': 'https://drive.google.com/uc?export=download&id=1I46pZG4Z8DECIEBXA2CYnkwzub1F13Tc',\n","        'rppa': 'https://drive.google.com/uc?export=download&id=1rL47G7dVaaQQrrntKLr3hkZKe_ZNmSap'\n","    },\n","    'prostate': {\n","        'clinical': 'https://drive.google.com/uc?export=download&id=1mwa5Jbmks7x07hNzyEVEHAtk61_Fx2wg',\n","        'methylation': 'https://drive.google.com/uc?export=download&id=1y_zN-dJddduH9kr7b6aHP0NWcrJM5IeP',\n","        'mirna': 'https://drive.google.com/uc?export=download&id=15eia1T6QOULnJtiob1KpEzSZGzNrjgrd',\n","        'rnaseq': 'https://drive.google.com/uc?export=download&id=178nuI9SpWVkO7-tSlKFTVLSMsR7FK8BP',\n","        'rppa': 'https://drive.google.com/uc?export=download&id=1It0DmCOtszQx6xgLhVHfen81rrt6AHhz'\n","    }\n","}\n","\n","# Function to download the file if it doesn't exist\n","def download_file(file_name, file_url):\n","    directory = os.path.dirname(file_name)\n","\n","    # Ensure directory exists\n","    if not os.path.exists(directory):\n","        os.makedirs(directory)\n","\n","    if not os.path.exists(file_name):\n","        print(f\"Downloading {file_name}...\")\n","        urllib.request.urlretrieve(file_url, file_name)\n","\n","# Function to load and download LinkedOmics data\n","def load_data(cancer_type, feature_types, response=None, remove_nan=True, correlation_threshold=0.999, see_names=False, verbose=False):\n","\n","    file_paths = {\n","        'clinical': f\"{cancer_type}/clinical.txt\",\n","        'methylation': f\"{cancer_type}/methylation.txt\",\n","        'mirna': f\"{cancer_type}/mirna.txt\",\n","        'rnaseq': f\"{cancer_type}/rnaseq.txt\",\n","        'rppa': f\"{cancer_type}/rppa.txt\"\n","    }\n","\n","    # Download the required files based on the selected cancer_type\n","    for feature in feature_types:\n","        download_file(file_paths[feature], file_urls[cancer_type][feature])\n","\n","    if response is not None:\n","        response_type, _ = response\n","        download_file(file_paths[response_type], file_urls[cancer_type][response_type])\n","\n","    # Initialize shared labels\n","    shared_labels = None\n","\n","    # Handle the response variable\n","    if response is not None:\n","        response_type, response_name = response\n","        response_df = load_dataframe(file_paths[response_type])\n","        Y_df = response_df[response_name].apply(pd.to_numeric, errors='coerce')\n","        shared_labels = set(Y_df.index)\n","    else:\n","        response_type = None\n","\n","    # Features\n","    feature_dfs = []\n","    feature_names = []\n","    for category in feature_types:\n","        df = load_dataframe(file_paths[category])\n","        if category == response_type:\n","            df = df.drop(columns=[response_name], errors='ignore')\n","        if shared_labels is not None:\n","            shared_labels = shared_labels.intersection(df.index)\n","        else:\n","            shared_labels = set(df.index)\n","        feature_dfs.append(df)\n","        feature_names.extend(df.columns.tolist())\n","        if category == 'mirna':\n","            feature_names = [mirna.replace('hsa-mir-', 'miR-') for mirna in feature_names]\n","            feature_names = [mirna.replace('hsa-let-', 'let-') for mirna in feature_names]\n","\n","    # Ensure shared labels exist\n","    if not shared_labels:\n","        raise ValueError(\"No shared labels found across the feature datasets.\")\n","\n","    shared_labels = sorted(shared_labels)\n","    feature_dfs = [df.loc[shared_labels] for df in feature_dfs]\n","\n","    if verbose:\n","        print(f'Number of rows with shared labels: {len(shared_labels)}')\n","\n","    if see_names:\n","        print(feature_names)\n","        return\n","\n","    # Concatenate features into X\n","    X = pd.concat(feature_dfs, axis=1).to_numpy()\n","    X = X.astype(float)\n","\n","    # Remove columns with NaN values\n","    if remove_nan:\n","        cols_with_nan = np.isnan(X).any(axis=0)\n","        X = X[:, ~cols_with_nan]\n","        feature_names = [name for (name, keep) in zip(feature_names, ~cols_with_nan) if keep]\n","\n","    # Remove highly correlated columns (correlation > correlation_threshold)\n","    if X.shape[1] > 1:  # Check if there is more than one column\n","        correlation_matrix = np.corrcoef(X, rowvar=False)\n","\n","        # Identify pairs of columns with correlation > threshold\n","        correlated_columns = set()\n","        for i in range(len(correlation_matrix)):\n","            for j in range(i + 1, len(correlation_matrix)):\n","                if abs(correlation_matrix[i, j]) > correlation_threshold:\n","                    correlated_columns.add(j)\n","\n","        # Keep uncorrelated columns\n","        uncorrelated_columns = [i for i in range(X.shape[1]) if i not in correlated_columns]\n","        X = X[:, uncorrelated_columns]\n","        feature_names = [feature_names[i] for i in uncorrelated_columns]\n","\n","        if verbose:\n","            print(f'Removed {len(correlated_columns)} columns due to high correlation.')\n","\n","    if response is not None:\n","        Y = Y_df.loc[shared_labels].to_numpy()\n","        rows_with_nan = np.isnan(Y)\n","        X = X[~rows_with_nan, :]\n","        Y = Y[~rows_with_nan]\n","        return {'X': X, 'Y': Y, 'feature_names': feature_names, 'response': response}\n","    else:\n","        return {'X': X, 'Y': None, 'feature_names': feature_names, 'response': response}\n","\n","# Function to load dataframes from local files\n","def load_dataframe(file_path):\n","    df = pd.read_csv(file_path, sep='\\t').T\n","    df.columns = df.iloc[0]\n","    df = df.drop(df.index[0])\n","    df.index.name = 'attrib_name'\n","    return df"],"metadata":{"id":"0vE3S8_aDXvc"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Available options for the different data types:\n","- cancer type: 'ovarian' or 'prostate'\n","- feature type: 'clinical', 'methylation', 'mirna', 'rnaseq', and/or 'rppa'\n","- response type: None if no response, else tuple of strings (feature_type, feature_name), where\n","\t- feature_type is any one of the feature types above\n","\t- feature_name is any feature name in the specified feature type, e.g., 'Tumor_purity' in 'clinical'\n","\n","For further details about the datasets:\n","- ovarian cancer: https://www.linkedomics.org/data_download/TCGA-OV/\n","- prostate cancer: https://www.linkedomics.org/data_download/TCGA-PRAD/"],"metadata":{"id":"-N8SE6YpEpkB"}},{"cell_type":"code","source":["# uncomment both lines below to print all feature names for a given feature type (e.g., to see response variable options)\n","# cancer_type, feature_types = 'ovarian', ['clinical']\n","# data = load_data(cancer_type, feature_types, see_names=True)\n","\n","cancer_type = 'prostate'\n","feature_types = ['mirna']\n","response = ('clinical', 'Tumor_purity')\n","\n","data = load_data(cancer_type, feature_types, response=response)\n","X, y, feature_names = data['X'], data['Y'], data['feature_names']\n","print(f'Number of samples = {X.shape[0]}')\n","print(f'Number of features = {X.shape[1]}')"],"metadata":{"id":"FrZH-AFOBLwp"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Run IPSS"],"metadata":{"id":"NFVLHrDwFCYD"}},{"cell_type":"code","source":["ipss_output = ipss(X, y, selector='gb')"],"metadata":{"id":"gtgT4nHdCGnb"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Analyze results"],"metadata":{"id":"T1Ro8IbEFEzb"}},{"cell_type":"code","source":["plot_q_values = True\n","plot_efp_scores = True\n","\n","# plot q-values for all features with q-values below a certain threshold\n","if plot_q_values:\n","\tq_value_threshold = 0.5\n","\tq_values = ipss_output['q_values']\n","\tq_values = [(index, q_value) for (index, q_value) in q_values if q_value <= q_value_threshold]\n","\n","\tplt.figure(figsize=(10, 6))\n","\tfor i, (feature_index, q_value) in enumerate(q_values):\n","\t\tplt.bar(i, q_value, color='dodgerblue')\n","\tplt.xticks(range(len(q_values)), [feature_names[feature_index] for feature_index, _ in q_values], rotation=45)\n","\tplt.ylabel('$q$-value', fontsize=18)\n","\tplt.tight_layout()\n","\tplt.show()\n","\n","# plot efp scores for all features with efp scores below a certain threshold\n","if plot_efp_scores:\n","\tefp_score_threshold = 5\n","\tefp_scores = ipss_output['efp_scores']\n","\tefp_scores = [(index, q_value) for (index, q_value) in efp_scores if q_value <= efp_score_threshold]\n","\n","\tplt.figure(figsize=(10, 6))\n","\tfor i, (feature_index, efp_score) in enumerate(efp_scores):\n","\t\tplt.bar(i, efp_score, color='dodgerblue')\n","\tplt.xticks(range(len(efp_scores)), [feature_names[feature_index] for feature_index, _ in efp_scores], rotation=45)\n","\tplt.ylabel('efp score', fontsize=18)\n","\tplt.tight_layout()\n","\tplt.show()"],"metadata":{"id":"H2aiW5bFCbtN"},"execution_count":null,"outputs":[]}]}